pykern/util.py

```diff
@@ -65,3 +70,48 @@ def _check_dev_files(root):
         # Don't run from an install directory
         r = pkio.py_path()
     return pkio.mkdir_parent(r.join(_DEFAULT_ROOT))
+
+
+def is_pure_text(value, test_size=512):
+    """Guesses if value is pure text using heuristics.
+
+    Args:
+        value (bytes): bytes data
+        test_size (int): size of bytes chunk being tested
+
+    Returns:
+        bool: True if bytes_data is likely pure text, false if likely binary
+    """
+
+    def _is_accepted_control_code_ratio(value):
+        c = 0
+        for i in range(len(value)):
+            if bytes([value[i]]) in _STRANGE_CONTROL_CODES:
+                c += 1
+        return (c / len(value)) < (1 / 3)
+
+    def _try(chunk):
+        try:
+            chunk.decode("utf-8", "strict")
+            return b"\x00" not in value
+        except UnicodeDecodeError:
+            return False
+
+    def _valid_unicode(value, test_size):
+        if len(value) <= test_size:
+            return _try(value)
+        b = value[:test_size]
+        # 4 is maximum length of a utf8 char so if a char
+        # is truncated by test_size, we need to probe back
+        # a bit to find the end of the char.
+        for _ in range(4):
+            if _try(b):
+                return True
+            if len(b) <= 1:
+                return False
+            b = b[:-1]
+        return False
+
+    if _valid_unicode(value, test_size):
+        return _is_accepted_control_code_ratio(value)
```

Comment:
This has to be on the decoded version of the str. You'll have to ord() each utf8 character. You might have a check if the ord is above 127 or something.